lib/runner.js

```diff
@@ -320,13 +320,18 @@ Runner.prototype.hook = function (name, fn) {
         } else {
           self.failHook(hook, err);
 
-          // stop executing hooks, notify callee of hook err
-          return fn(err);
+          if (name === 'beforeAll' || name === 'beforeEach') {
+            // stop executing hooks, notify callee of hook err
+            return fn(err);
+          } else {
+            // continue calling all after and afterEach hooks
+            prevErr = err;
+          }
         }
       }
       self.emit('hook end', hook);
       delete hook.ctx.currentTest;
-      next(++i);
+      next(++i, prevErr);
```

Comment:
so I understand: we're trapping the first error we find, then passing it along to the rest of the `after` and `afterEach` hooks?  then, if there's no hooks remaining, we finally bail?

Reply:
yes

The prevErr is just to keep track of the last error that happened. The first after hook might throw an error, and the next ones succeed. Then the error has to be remembered so it can be used when calling the callback fn.